File: src/scripts/token_bucket.lua
-- TOKEN BUCKET ALGORITHM (Redis Lua Script)
-- Atomic execution prevents race conditions in concurrent environments
--
-- How it works:
-- A bucket holds a max number of tokens. Each request consumes 1 token.
-- Tokens refill at a steady rate over time (e.g., 100 tokens per minute).
-- If the bucket is empty, the request is rejected.
--
-- KEYS[1] = rate limit key (e.g., "rl:token_bucket:user:123")
-- ARGV[1] = max tokens (bucket capacity)
-- ARGV[2] = refill rate (tokens per second)
-- ARGV[3] = current timestamp (milliseconds)
-- ARGV[4] = tokens requested (usually 1)

local key = KEYS[1]
local max_tokens = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])  -- tokens per second
local now = tonumber(ARGV[3])
local requested = tonumber(ARGV[4]) or 1

-- Retrieve current bucket state from Redis
local bucket = redis.call('HMGET', key, 'tokens', 'last_refill')
local tokens = tonumber(bucket[1])
local last_refill = tonumber(bucket[2])

-- First request for this key: initialize a full bucket
if tokens == nil then
  tokens = max_tokens
  last_refill = now
end

-- Calculate how many tokens to add based on elapsed time
local elapsed_seconds = (now - last_refill) / 1000
local tokens_to_add = elapsed_seconds * refill_rate

-- Refill the bucket (but don't exceed max capacity)
tokens = math.min(max_tokens, tokens + tokens_to_add)
last_refill = now

-- Check if enough tokens exist to serve the request
local allowed = 0
local remaining = 0

if tokens >= requested then
  tokens = tokens - requested
  allowed = 1
  remaining = math.floor(tokens)
else
  allowed = 0
  remaining = 0
end

-- Persist updated state with TTL (auto-cleanup after inactivity)
local ttl = math.ceil(max_tokens / refill_rate) * 2
redis.call('HMSET', key, 'tokens', tokens, 'last_refill', last_refill)
redis.call('EXPIRE', key, ttl)

-- Return: [allowed (0/1), remaining_tokens, retry_after_ms]
local retry_after = 0
if allowed == 0 then
  retry_after = math.ceil((requested - tokens) / refill_rate * 1000)
end

return {allowed, remaining, retry_after, math.floor(max_tokens)}

File: src/algorithms/token-bucket.js
/**
 * TOKEN BUCKET ALGORITHM
 * 
 * Concept: Imagine a bucket that can hold N tokens.
 * - Tokens are added at a fixed rate (e.g., 10/sec)
 * - Each request consumes 1 token
 * - If the bucket is full, new tokens are discarded
 * - If the bucket is empty, requests are rejected
 * 
 * Key property: Allows short bursts (if bucket has accumulated tokens)
 * while enforcing an average rate limit over time.
 * 
 * Example: maxTokens=100, refillRate=10/sec
 * - A user can burst 100 requests immediately
 * - Then must wait for tokens to refill at 10/sec
 * - Average throughput is capped at 10 req/sec
 */

const redisClient = require('../redis-client');
const { v4: uuidv4 } = require('uuid');

class TokenBucketLimiter {
  /**
   * @param {object} options
   * @param {number} options.maxTokens     - Bucket capacity (burst limit)
   * @param {number} options.refillRate    - Tokens added per second
   * @param {string} options.keyPrefix     - Redis key namespace
   */
  constructor(options = {}) {
    this.maxTokens = options.maxTokens || 100;
    this.refillRate = options.refillRate || 10;
    this.keyPrefix = options.keyPrefix || 'rl:token_bucket';
  }

  /**
   * Check if a request is allowed under the token bucket algorithm.
   * 
   * @param {string} identifier - Unique identifier (user ID, IP, API key)
   * @param {number} tokensRequested - Tokens to consume (default: 1)
   * @returns {object} { allowed, remaining, retryAfter, limit }
   */
  async isAllowed(identifier, tokensRequested = 1) {
    const key = `${this.keyPrefix}:${identifier}`;
    const now = Date.now();

    const result = await redisClient.evalScript('tokenBucket', [key], [
      this.maxTokens,
      this.refillRate,
      now,
      tokensRequested,
    ]);

    // Lua returns: [allowed, remaining, retry_after_ms, max_tokens]
    const [allowed, remaining, retryAfter, limit] = result;

    return {
      allowed: allowed === 1,
      remaining: Math.floor(remaining),
      retryAfter: retryAfter > 0 ? Math.ceil(retryAfter / 1000) : 0, // convert to seconds
      limit: limit,
      algorithm: 'token_bucket',
      identifier,
    };
  }

  /**
   * Get the current bucket state without consuming a token.
   * Useful for monitoring dashboards.
   */
  async getStatus(identifier) {
    const key = `${this.keyPrefix}:${identifier}`;
    const client = redisClient.getClient();
    const bucket = await client.hmget(key, 'tokens', 'last_refill');

    return {
      tokens: parseFloat(bucket[0]) || this.maxTokens,
      lastRefill: parseInt(bucket[1]) || Date.now(),
      maxTokens: this.maxTokens,
      refillRate: this.refillRate,
    };
  }

  /**
   * Reset a user's bucket (e.g., for admin override or testing)
   */
  async reset(identifier) {
    const key = `${this.keyPrefix}:${identifier}`;
    await redisClient.getClient().del(key);
    return { reset: true, identifier };
  }
}

module.exports = TokenBucketLimiter;
